README.rst
=====
HG502
=====

HG502 is an application with a graphical user interface that allows you
to track your progress in the Holy Grail 502 Diablo 2 challenge.

Supported files:

- .d2s - Diablo 2 save files;
- .d2x - PlugY personal stash files;
- .sss - PlugY shared stash files.

Features:

- statistics for each item type such as set's and unique items and overall;
- ability to view lists of found and remaining items for each type
  and overall;
- search through lists of items.

------------
Installation
------------
- download the archive for Windows and Linux that does not require installation `here <https://github.com/artcom-net/hg502/releases>`_;
- extract the archive to a convenient directory;
- run HG502.exe for Windows or HG502 for Linux.

-----
Usage
-----
It is easy to use just specify the folder with Diablo 2 saves: Open - > Folder.

.. image:: images/main_win.png

CHANGELOG.md
# Changelog

## [0.1.1](https://github.com/artcom-net/hg502/tree/v0.1.1) (2020-02-16)

### Fixed
- exclude "Standard of Heroes" from the list of unique items
- items from a corpse (if the character is dead) and a mercenary are counted

tests/test_hg502.py
import json
from pathlib import Path

import pytest

from hg502_tracker.hg502 import HG502

SAVE_PATH = 'data'
ITEMS_DICT = {0: 'Test0', 1: 'Test1', 2: 'Test2', 3: 'Test3'}
USER_ITEMS_DICT = {4: 'Test4', 5: 'Test5'}


@pytest.fixture(scope='module')
def hg502():
    return HG502()


@pytest.fixture(scope='module')
def hg502_expected(hg502):
    set_stat = None
    unique_stat = None
    total_stat = None
    for path in Path(SAVE_PATH).glob('*.json'):
        if path.stem == 'total_stat':
            total_stat = json.loads(path.open('r').read())
        elif path.stem == 'set_stat':
            set_stat = json.loads(path.open('r').read())
        elif path.stem == 'unique_stat':
            unique_stat = json.loads(path.open('r').read())
    return hg502, total_stat, set_stat, unique_stat


def test_hg502_constructor(hg502):
    assert hg502._set_dict
    assert hg502._unique_dict
    assert len(hg502._set_dict) == 127
    assert len(hg502._unique_dict) == 375

    for quest_item_id in hg502._QUESTS_UNIQUE:
        assert quest_item_id not in hg502._unique_dict
    for die_facet_id, level_up_facet_id in hg502._FACET_PAIRS:
        assert die_facet_id in hg502._unique_dict
        assert level_up_facet_id not in hg502._unique_dict

    assert hg502._save_path is None
    assert not hg502._user_set_items
    assert not hg502._user_unique_items
    assert isinstance(hg502._user_set_items, dict)
    assert isinstance(hg502._user_unique_items, dict)


def test_hg502_get_stat_dict(hg502):
    stat_dict = hg502._get_stat_dict()
    assert stat_dict
    assert isinstance(stat_dict, dict)
    assert stat_dict['total_items'] == 0
    assert stat_dict['total_found'] == 0
    assert stat_dict['total_remaining'] == 0
    assert stat_dict['progress'] == 0.0
    assert isinstance(stat_dict['found_items'], list)
    assert isinstance(stat_dict['remaining_items'], list)


@pytest.mark.parametrize(
    'total,part,expected',
    ((150, 15, 10.0), (100, 100, 100.0), (502, 348, 69.32270916334662)),
)
def test_hg502_calc_percentage(hg502, total, part, expected):
    assert hg502._calc_percentage(total, part) == expected


def test_hg502_get_hg502_stat(hg502_expected):
    hg502, total_stat_exp, set_stat_exp, unique_stat_exp = hg502_expected
    total_stat, set_stat, unique_stat = hg502.get_hg502_stat(SAVE_PATH)
    assert total_stat == total_stat_exp
    assert set_stat == set_stat_exp
    assert unique_stat == unique_stat_exp


@pytest.mark.parametrize(
    'stat_dict,items_dict,user_items_dict,expected',
    (
        (
            {
                'total_items': 0,
                'total_found': 0,
                'total_remaining': 0,
                'progress': 0.0,
                'found_items': [],
                'remaining_items': [],
            },
            ITEMS_DICT,
            USER_ITEMS_DICT,
            {
                'total_items': 4,
                'total_found': 2,
                'total_remaining': 2,
                'progress': 50.0,
                'found_items': ['Test4', 'Test5'],
                'remaining_items': [],
            },
        ),
    ),
)
def test_hg502_get_common_stat(
    hg502, stat_dict, items_dict, user_items_dict, expected
):
    hg502._get_common_stat(stat_dict, items_dict, user_items_dict)
    assert stat_dict == expected


@pytest.mark.parametrize(
    'items_dict,user_items_dict,expected',
    ((ITEMS_DICT, USER_ITEMS_DICT, {0, 1, 2, 3}),),
)
def test_hg502_get_remaining_ids(hg502, items_dict, user_items_dict, expected):
    assert hg502._get_remaining_ids(items_dict, user_items_dict) == expected


def test_hg502_load_user_items(hg502):
    hg502._load_user_items(SAVE_PATH)
    assert hg502._save_path == SAVE_PATH
    assert hg502._user_set_items
    assert hg502._user_unique_items


def test_hg502_load_user_items_invalid_path(hg502):
    with pytest.raises(FileNotFoundError):
        hg502._load_user_items('.')


@pytest.mark.parametrize(
    'facet_id,expected', ((392, True), (399, True), (391, False), (400, False))
)
def test_hg502_is_facet(hg502, facet_id, expected):
    assert hg502._is_facet(facet_id) is expected


@pytest.mark.parametrize(
    'facet_id,expected', ((392, 392), (397, 393), (399, 395), (394, 394))
)
def test_hg502_get_die_facet_id(hg502, facet_id, expected):
    assert hg502._get_die_facet_id(facet_id) == expected


@pytest.mark.parametrize(
    'facet_id,expected',
    ((392, 'Lightning'), (398, 'Fire'), (399, 'Poison'), (393, 'Cold')),
)
def test_hg502_get_facet_suffix(hg502, facet_id, expected):
    assert hg502._get_facet_suffix(facet_id) == expected


build.py
import os
import site
import sys
from pathlib import Path
from tarfile import TarFile
from zipfile import ZIP_DEFLATED, ZipFile

import PyInstaller.__main__

from hg502_tracker import __app_name__, __version__

BUILD_DIR = 'build'
DIST_DIR = 'dist'
UI_DIR = 'ui'
STATIC_DIR = 'static'
ICON_NAME = 'icon.ico'
SCRIPT_PATH = 'hg502_tracker/main.py'


def retrieve_file_paths(directory):
    file_list = []
    for root, dirs, files in os.walk(directory):
        for file in files:
            file_list.append(Path(root).joinpath(file))

    return file_list


def build_package():
    build_path = Path(BUILD_DIR)
    build_path.mkdir(exist_ok=True)
    work_path = build_path.joinpath('build')
    build_dist_path = build_path.joinpath('dist')

    site_packages_path = None
    for path in site.getsitepackages():
        if Path(path).stem == 'site-packages':
            site_packages_path = path

    items_data = 'd2lib/items_data'
    items_data_path = Path(site_packages_path).joinpath(items_data)

    ui_path = Path(UI_DIR).absolute()
    static_path = Path(STATIC_DIR).absolute()
    icon_path = static_path.joinpath(ICON_NAME).absolute()

    path_sep = ';' if sys.platform == 'win32' else ':'

    _items_data = f'{items_data_path}{path_sep}{items_data}'
    ui_data = f'{ui_path}{path_sep}{UI_DIR}'
    static_data = f'{static_path}{path_sep}{STATIC_DIR}'

    PyInstaller.__main__.run(
        [
            '--clean',
            '--noconfirm',
            '--onedir',
            '--windowed',
            f'--name={__app_name__}',
            f'--workpath={work_path}',
            f'--distpath={build_dist_path}',
            f'--specpath={work_path}',
            f'--add-data={_items_data}',
            f'--add-data={ui_data}',
            f'--add-data={static_data}',
            f'--icon={icon_path}',
            SCRIPT_PATH,
        ]
    )

    file_list = retrieve_file_paths(build_dist_path.joinpath(__app_name__))
    out_file_name = '-'.join(
        (f'{__app_name__}', f'{__version__}', f'{sys.platform}',)
    )
    dist_path = Path(DIST_DIR)
    dist_path.mkdir(exist_ok=True)
    archive_path = dist_path.joinpath(f'{out_file_name}')

    if sys.platform == 'win32':
        zip_file = ZipFile(f'{archive_path}.zip', 'w', ZIP_DEFLATED)
        with zip_file:
            for file_path in file_list:
                zip_file.write(
                    file_path, arcname=file_path.relative_to(build_dist_path)
                )
    else:
        tar_file = TarFile.open(f'{archive_path}.tar.gz', 'w:gz')
        with tar_file:
            for file_path in file_list:
                tar_file.add(
                    file_path, arcname=file_path.relative_to(build_dist_path)
                )


if __name__ == '__main__':
    build_package()


hg502_tracker/app.py
import sys
from pathlib import Path

import qdarkstyle

from hg502_tracker import (
    __app_name__,
    __author__,
    __author_email__,
    __home_page__,
    __license__,
    __version__,
)
from hg502_tracker.hg502 import FileParseError


class HG502App(object):
    """This class is a presenter. Manages the backend and GUI."""

    def __init__(self, gui, backend, q_app):
        """Initializes an instance.

        :type gui: src.gui.HG502GUI
        :type backend: src.hg502_tracker.HG502
        :type q_app: QApplication
        """
        self._gui = gui
        self._backend = backend
        self._q_app = q_app
        self._home_path = Path().home()
        self._settings_path = self._home_path.joinpath('.hg502')
        self._save_path = None

        self._q_app.setStyleSheet(qdarkstyle.load_stylesheet())
        self._gui.set_about_data(
            __app_name__,
            __version__,
            __license__,
            __author__,
            __author_email__,
            f'<html><head/><body><p>'
            f'<a href="{__home_page__}">'
            f'<span style=" text-decoration: underline; color:#ffffff;">'
            f'{__home_page__}</span></a></p></body></html>',
        )
        self._gui.action_handler_register(
            '_action_folder', self._open_folder_handler
        )
        self._gui.action_handler_register('_action_exit', self._exit_handler)
        self._gui.clicked_handler_register(
            '_refresh_button', self._refresh_handler
        )

        if self._settings_path.exists():
            self._save_path = self._settings_path.read_text()
            self._show_stat()

    @staticmethod
    def _prepare_stat(stat_dict):
        """Prepares statistics data for the GUI.

        :type stat_dict: dict
        :rtype: dict
        """
        stat_fields = ('total_items', 'total_found', 'total_remaining')
        stat = [
            str(stat_dict[field])
            for field in stat_dict
            if field in stat_fields
        ]
        stat.append(f'{stat_dict["progress"]:.2f}%')
        return stat

    def _show_stat(self):
        """Displays statistics in a GUI.

        :rtype: bool
        """
        if not self._push_stat():
            return False
        self._gui.set_status_bar_text(f'Current folder: {self._save_path}')
        if not self._gui.is_enabled_widgets:
            self._gui.set_widgets_status('enable')
        return True

    def _push_stat(self):
        """Fills widgets with necessary data.

        :rtype: bool
        """
        try:
            total_stat, set_stat, unique_stat = self._backend.get_hg502_stat(
                self._save_path
            )
        except FileNotFoundError:
            self._gui.show_info_message(
                'Info',
                'Files not found',
                f'{self._save_path} does not contain Diablo 2 files',
            )
            return False
        except FileParseError as err:
            self._gui.show_error_message('Error', 'File parse error', f'{err}')
            return False

        stats = []
        for stat_dict in (set_stat, unique_stat, total_stat):
            stats.append(self._prepare_stat(stat_dict))
        self._gui.fill_stat_table(stats)
        self._gui.fill_set_items_list(
            set_stat['found_items'], set_stat['remaining_items']
        )
        self._gui.fill_unique_items_list(
            unique_stat['found_items'], unique_stat['remaining_items']
        )

        all_found = set_stat['found_items'].copy()
        all_found.extend(unique_stat['found_items'].copy())
        all_remaining = set_stat['remaining_items'].copy()
        all_remaining.extend(unique_stat['remaining_items'].copy())

        self._gui.fill_all_items_list(all_found, all_remaining)

        return True

    def _open_folder_handler(self):
        """`Folder...` button event handler.

        Displays the explorer for directory selection and calls _show_stat
        if input was received.
        """
        folder_path = self._gui.show_open_folder_dialog(
            'Diablo 2 save folder', directory=str(self._home_path)
        )
        if folder_path:
            self._save_path = folder_path
            if self._show_stat():
                self._settings_path.write_text(self._save_path)

    def _refresh_handler(self):
        """`Refresh` button event handler."""
        self._push_stat()

    def _exit_handler(self):
        """`Exit` button event handler."""
        self._q_app.quit()

    def run(self):
        """Displays GUI."""
        self._gui.show()
        sys.exit(self._q_app.exec_())


hg502_tracker/gui.py
import sys
from pathlib import Path

from PyQt5 import QtCore, uic
from PyQt5.QtWidgets import (
    QAbstractItemView,
    QFileDialog,
    QLabel,
    QListWidget,
    QMainWindow,
    QMessageBox,
    QTableWidgetItem,
)

from hg502_tracker import __app_name__

UI_DIR = 'ui'


def get_resource_path(relative_path):
    """Get absolute path to resource, works for dev and for PyInstaller.

    PyInstaller creates a temp folder and stores path in _MEIPASS.

    :type relative_path: str
    :rtype: str
    """
    if hasattr(sys, '_MEIPASS'):
        return str(Path(sys._MEIPASS).joinpath(relative_path))
    return str(Path('..').joinpath(relative_path))


class _UILoader(object):
    """The mixin class implements the loading of GUI parameters from *.ui."""

    _UI_FILE = None

    def __init__(self):
        """Initializes an instance."""
        uic.loadUi(
            Path(get_resource_path(UI_DIR)).joinpath(self._UI_FILE), self
        )


class _AboutWindow(QMainWindow, _UILoader):
    """About window class."""

    _UI_FILE = 'about_window.ui'

    def __init__(self):
        """Initializes an instance."""
        super(_AboutWindow, self).__init__()


class HG502GUI(QMainWindow, _UILoader):
    """Class of the main interface window."""

    _UI_FILE = 'main_window.ui'

    def __init__(self):
        """Initializes an instance."""
        super(HG502GUI, self).__init__()
        self.setWindowTitle(__app_name__)

        self.is_enabled_widgets = False
        self._about = _AboutWindow()

        self._status_label = QLabel()
        self._status_bar.addWidget(self._status_label)

        self._action_about.triggered.connect(self._show_about)
        self._search_button.clicked.connect(self._search_handler)
        self._items_tab.currentChanged.connect(self._change_tab_handler)

        self.set_widgets_status('disable')

    def _get_list_widgets(self):
        """Returns all QListWidget from the current tab."""
        return self._items_tab.currentWidget().findChildren(QListWidget)

    def _change_tab_handler(self):
        """Scrolls up when the tab is switched."""
        for list_w in self._get_list_widgets():
            list_w.scrollToTop()

    def _show_about(self):
        """Shows the about window."""
        self._about.show()

    def closeEvent(self, *args, **kwargs):
        """Closes the about window if the user has closed the main window."""
        if self._about.isVisible():
            self._about.close()

    def set_about_data(
        self, app_name, version, _license, author, author_email, home_page
    ):
        """Sets the values for the QLabel of the about window.

        :type app_name: str
        :type version: str
        :type _license: str
        :type author: str
        :type author_email: str
        :type home_page: str
        """
        self._about._app_name_label.setText(app_name)
        self._about._version_value.setText(version)
        self._about._license_value.setText(_license)
        self._about._author_value.setText(author)
        self._about._author_email_value.setText(author_email)
        self._about._home_page_value.setText(home_page)

    def action_handler_register(self, action, func):
        """Registers a handler for an action.

        :type action: str
        :type func: function
        """
        _action = getattr(self, action)
        _action.triggered.connect(func)

    def clicked_handler_register(self, widget, func):
        """Registers a handler for a clicked event.

        :type widget: str
        :type func: function
        """
        obj = getattr(self, widget)
        obj.clicked.connect(func)

    def show_open_folder_dialog(self, title, directory):
        """Displays the directory selection dialog.

        :param title: Window title
        :type title: str
        :param directory: Directory to open
        :type directory: str
        :return: Path to the directory
        :rtype: str
        """
        return QFileDialog.getExistingDirectory(
            self, title, directory=directory
        )

    @staticmethod
    def _clear_selected(list_widget):
        """Resets selection from QListWidget elements.

        :type list_widget: QListWidget
        """
        for item in list_widget.selectedItems():
            item.setSelected(False)

    def _search_handler(self):
        """Handler for the search button click event."""
        search_text = self._search_line.text().strip()
        if not search_text:
            return None
        for list_w in self._get_list_widgets():
            items = list_w.findItems(search_text, QtCore.Qt.MatchContains)
            self._clear_selected(list_w)
            list_w.setSelectionMode(QAbstractItemView.MultiSelection)
            if items:
                for item in items:
                    item.setSelected(True)
                    list_w.scrollToItem(
                        item, QAbstractItemView.PositionAtCenter
                    )
            list_w.setSelectionMode(QAbstractItemView.SingleSelection)

    def set_widgets_status(self, status):
        """Enables / disables widgets.

        :param status: 'enabled' or 'disabled'
        :type status: str
        """
        is_enable = True if status == 'enable' else False
        for widget in (
            self._stat_table,
            self._refresh_button,
            self._search_line,
            self._search_button,
            self._items_tab,
        ):
            widget.setEnabled(is_enable)
        self.is_enabled_widgets = is_enable

    def fill_stat_table(self, data):
        """Fills in the statistics table.

        :param data: A list of three elements where each element is 4 long and
        is a table row.
        :type data: iterable
        """
        row = 0
        for row_data in data:
            column = 0
            for value in row_data:
                item = QTableWidgetItem(str(value))
                item.setTextAlignment(QtCore.Qt.AlignCenter)
                self._stat_table.setItem(row, column, item)
                column += 1
            row += 1

    @staticmethod
    def _fill_items_list(w_list, data):
        """Fills the QListWidget with data from `data`.

        :type w_list: QListWidget
        :type data: list
        """
        w_list.clear()
        for value in data:
            w_list.addItem(value)
        w_list.scrollToTop()

    def fill_all_items_list(self, found, remaining):
        """Fills the QListWidget of the `All` tab.

        :type found: list
        :type remaining: list
        """
        self._fill_items_list(self._all_list_found, found)
        self._fill_items_list(self._all_list_remaining, remaining)

    def fill_set_items_list(self, found, remaining):
        """Fills the QListWidget of the `Set's` tab.

        :type found: list
        :type remaining: list
        """
        self._fill_items_list(self._set_list_found, found)
        self._fill_items_list(self._set_list_remaining, remaining)

    def fill_unique_items_list(self, found, remaining):
        """Fills the QListWidget of the `Unique` tab.

        :type found: list
        :type remaining: list
        """
        self._fill_items_list(self._unique_list_found, found)
        self._fill_items_list(self._unique_list_remaining, remaining)

    @staticmethod
    def _init_message(msg_type, title, short_text, info_text):
        """Initializes message.

        :type msg_type: QMessageBox.Information | QMessageBox.Critical
        :type title: str
        :type short_text: str
        :type info_text: str
        :rtype: QMessageBox
        """
        msg = QMessageBox()
        msg.setIcon(msg_type)
        msg.setText(short_text)
        msg.setInformativeText(info_text)
        msg.setWindowTitle(title)
        return msg

    def show_info_message(self, title, short_text, info_text):
        """Displays an informational message.

        :type title: str
        :type short_text: str
        :type info_text: str
        """
        msg = self._init_message(
            QMessageBox.Information, title, short_text, info_text
        )
        msg.exec_()

    def show_error_message(self, title, short_text, error_text):
        """Displays an error message.

        :type title: str
        :type short_text: str
        :type error_text: str
        """
        msg = self._init_message(
            QMessageBox.Critical, title, short_text, error_text
        )
        msg.exec_()

    def set_status_bar_text(self, text):
        """Sets the value of the status bar.

        :type text: str
        """
        self._status_label.setText(text)


hg502_tracker/hg502.py
from pathlib import Path

from d2lib.errors import D2SFileParseError, ItemParseError, StashFileParseError
from d2lib.files import D2SFile, D2XFile, SSSFile
from d2lib.items_storage import ItemsDataStorage


class FileParseError(Exception):
    """Used in case of errors in parsing user files."""

    pass


class HG502(object):
    """This class retrieves user item data."""

    _QUESTS_UNIQUE = (123, 124, 125, 126, 127, 128, 4095)
    # (die_facet, level_up_facet)
    _FACET_LIGHT = (392, 396)
    _FACET_COLD = (393, 397)
    _FACET_FIRE = (394, 398)
    _FACET_POISON = (395, 399)
    _FACET_SUFFIXES = {
        _FACET_LIGHT: 'Lightning',
        _FACET_COLD: 'Cold',
        _FACET_FIRE: 'Fire',
        _FACET_POISON: 'Poison',
    }
    _FACET_PAIRS = (_FACET_LIGHT, _FACET_COLD, _FACET_FIRE, _FACET_POISON)

    def __init__(self):
        """Initializes an instance."""
        item_storage = ItemsDataStorage()
        self._set_dict = item_storage.get_set_dict()
        self._unique_dict = item_storage.get_unique_dict()

        for quest_item_id in self._QUESTS_UNIQUE:
            self._unique_dict.pop(quest_item_id)
        for _, level_up_facet_id in self._FACET_PAIRS:
            self._unique_dict.pop(level_up_facet_id)

        self._save_path = None
        self._user_set_items = {}
        self._user_unique_items = {}

    @staticmethod
    def _get_stat_dict():
        """Returns the prepared dictionary for statistics."""
        return {
            'total_items': 0,
            'total_found': 0,
            'total_remaining': 0,
            'progress': 0.0,
            'found_items': [],
            'remaining_items': [],
        }

    @staticmethod
    def _calc_percentage(total, part):
        """Calculates the percentage of part from total.

        :type total: int
        :type part: int
        :rtype: float
        """
        return 100 * (part / total)

    def get_hg502_stat(self, save_path):
        """Collects statistics for all types of items.

        :param save_path: Path to Diablo 2 save directory
        :type save_path: str
        :return: Three dictionaries that look like this:
        {
            'total_items': int,
            'total_found': int,
            'total_remaining': int,
            'progress': float,
            'found_items': list,
            'remaining_items': list
        }
        :rtype: tuple
        """
        self._user_set_items.clear()
        self._user_unique_items.clear()

        total_stat = self._get_stat_dict()
        set_stat = self._get_stat_dict()
        unique_stat = self._get_stat_dict()
        self._load_user_items(save_path)

        if self._user_set_items:
            self._get_set_stat(set_stat)
        if self._user_unique_items:
            self._get_unique_stat(unique_stat)

        for sum_field in ('total_items', 'total_found', 'total_remaining'):
            total_stat[sum_field] += (
                set_stat[sum_field] + unique_stat[sum_field]
            )

        total_stat['progress'] = self._calc_percentage(
            total_stat['total_items'], total_stat['total_found']
        )

        for sorted_field in ('found_items', 'remaining_items'):
            set_stat[sorted_field].sort()
            unique_stat[sorted_field].sort()

        return total_stat, set_stat, unique_stat

    def _get_common_stat(self, stat_dict, items_dict, user_items_dict):
        """Fills in the general statistics for each type.

        :param stat_dict: Dictionary which is filled with data
        :type stat_dict: dict
        :param items_dict: Dictionary containing all items
        :type items_dict: dict
        :param user_items_dict: Dictionary with user items
        :type user_items_dict: dict
        """
        items_dict_len = len(items_dict)
        user_items_len = len(user_items_dict)
        stat_dict['total_items'] = items_dict_len
        stat_dict['total_found'] = user_items_len
        stat_dict['total_remaining'] = items_dict_len - user_items_len
        stat_dict['progress'] = self._calc_percentage(
            items_dict_len, user_items_len
        )
        stat_dict['found_items'] = list(user_items_dict.values())

    @staticmethod
    def _get_remaining_ids(items_dict, user_items_dict):
        """Calculates which item IDs the user does not have.

        :param items_dict: Dictionary containing all items
        :type items_dict: dict
        :param user_items_dict:  Dictionary with user items
        :type user_items_dict: dict
        :rtype: set
        """
        return set(items_dict).difference(set(user_items_dict))

    def _get_set_stat(self, stat_dict):
        """Collects statistics on set's items.

        :param stat_dict: Dictionary which is filled with data
        :type stat_dict: dict
        """
        self._get_common_stat(stat_dict, self._set_dict, self._user_set_items)
        remaining_items = []
        for remaining_id in self._get_remaining_ids(
            self._set_dict, self._user_set_items
        ):
            remaining_items.append(self._set_dict.get(remaining_id))

        stat_dict['remaining_items'] = remaining_items

    def _get_unique_stat(self, stat_dict):
        """Collects statistics on unique items.

        :param stat_dict: Dictionary which is filled with data
        :type stat_dict: dict
        """
        self._get_common_stat(
            stat_dict, self._unique_dict, self._user_unique_items
        )
        remaining_items = []
        for remaining_id in self._get_remaining_ids(
            self._unique_dict, self._user_unique_items
        ):
            if self._is_facet(remaining_id):
                remaining_items.append(
                    f'{self._unique_dict.get(remaining_id)} '
                    f'{self._get_facet_suffix(remaining_id)}'
                )
            else:
                remaining_items.append(self._unique_dict.get(remaining_id))

        stat_dict['remaining_items'] = remaining_items

    def _load_user_items(self, save_path):
        """Retrieves data for user items such as set's items and unique items.

        Data is taken from .d2s, .d2s and .sss files where .d2x is a PlugY
        personal stash file.

        :param save_path: Path to Diablo 2 save directory
        :type save_path: str
        :raises FileParseError:
        """
        self._save_path = save_path
        d2s_files = []
        stash_files = []

        try:
            for path in Path(self._save_path).iterdir():
                if path.suffix == '.d2s':
                    d2s_files.append(D2SFile(str(path)))
                elif path.suffix == '.d2x':
                    stash_files.append(D2XFile(str(path)))
                elif path.suffix == '.sss':
                    stash_files.append(SSSFile(str(path)))
        except (D2SFileParseError, StashFileParseError, ItemParseError) as err:
            raise FileParseError(f'{path}: {err}')

        if not any((d2s_files, stash_files)):
            raise FileNotFoundError

        for d2s in d2s_files:
            if d2s.items:
                self._filter_items(d2s.items)
            if d2s.corpse_items:
                self._filter_items(d2s.corpse_items)
            if d2s.merc_items:
                self._filter_items(d2s.merc_items)

        for stash_file in stash_files:
            for page in stash_file.stash:
                self._filter_items(page['items'])

    def _is_facet(self, item_id):
        """Returns True if the ID belongs to Rainbow facet otherwise False.

        :type item_id: int
        :rtype: bool
        """
        for pair in self._FACET_PAIRS:
            if item_id in pair:
                return True
        return False

    def _get_die_facet_id(self, facet_id):
        """Returns the ID of die facet by its pair.

        :type facet_id:
        :rtype: int
        """
        for pair in self._FACET_PAIRS:
            if facet_id in pair:
                return pair[0]

    def _get_facet_suffix(self, facet_id):
        """Returns an appropriate suffix by facet ID.

        :type facet_id: int
        :rtype: str
        """
        for pair, suffix in self._FACET_SUFFIXES.items():
            if facet_id in pair:
                return suffix

    def _filter_items(self, items):
        """Filters the desired items.

        For the HG 502 challenge, need set's and unique items. Rainbow facets
        count as four.

        :type items: list
        """
        for item in items:
            if item.is_set and item.set_id not in self._user_set_items:
                self._user_set_items[item.set_id] = item.name
            elif (
                item.is_unique
                and item.unique_id not in self._QUESTS_UNIQUE  # noqa
                and item.unique_id not in self._user_unique_items  # noqa
            ):
                if self._is_facet(item.unique_id):
                    die_facet_id = self._get_die_facet_id(item.unique_id)
                    if die_facet_id in self._user_unique_items:
                        continue
                    facet_name = (
                        f'{item.name} {self._get_facet_suffix(die_facet_id)}'
                    )
                    self._user_unique_items[die_facet_id] = facet_name
                    continue
                self._user_unique_items[item.unique_id] = item.name

            if item.socketed_items:
                self._filter_items(item.socketed_items)
